core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java

```diff
@@ -19,50 +19,120 @@
 
 package org.elasticsearch.index.translog;
 
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.index.IndexFormatTooNewException;
+import org.apache.lucene.index.IndexFormatTooOldException;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.apache.lucene.store.InputStreamDataInput;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.RamUsageEstimator;
 import org.elasticsearch.common.io.Channels;
+import org.elasticsearch.common.io.stream.InputStreamStreamInput;
 
+import java.io.Closeable;
 import java.io.EOFException;
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
  * a translog reader which is fixed in length
  */
-public class ImmutableTranslogReader extends TranslogReader {
+public class ImmutableTranslogReader extends TranslogReader implements Closeable {
+    private static final byte LUCENE_CODEC_HEADER_BYTE = 0x3f;
+    private static final byte UNVERSIONED_TRANSLOG_HEADER_BYTE = 0x00;
+
 
     private final int totalOperations;
     protected final long length;
+    protected final AtomicBoolean closed = new AtomicBoolean(false);
 
     /**
      * Create a snapshot of translog file channel. The length parameter should be consistent with totalOperations and point
      * at the end of the last operation in this snapshot.
      */
-    public ImmutableTranslogReader(long generation, ChannelReference channelReference, long firstOperationOffset, long length, int totalOperations) {
-        super(generation, channelReference, firstOperationOffset);
+    public ImmutableTranslogReader(long generation, FileChannel channel, Path path, long firstOperationOffset, long length, int totalOperations) {
+        super(generation, channel, path, firstOperationOffset);
         this.length = length;
         this.totalOperations = totalOperations;
     }
 
-    @Override
-    public final TranslogReader clone() {
-        if (channelReference.tryIncRef()) {
-            try {
-                ImmutableTranslogReader reader = newReader(generation, channelReference, firstOperationOffset, length, totalOperations);
-                channelReference.incRef(); // for the new object
-                return reader;
-            } finally {
-                channelReference.decRef();
+    /**
+     * Given a file, return a VersionedTranslogStream based on an
+     * optionally-existing header in the file. If the file does not exist, or
```

Comment:
This javadoc is stale?  The translog must exist, be non-zero length, have a proper header in 3.0?